### ISSUE TITLE
Feature suggestion: extensions to igraph_path_length_hist() for weighted and vertex-subset case

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#919]
This is a feature request to provide some extensions of the `igraph_path_length_hist()` functionality (possibly implemented as a separate function):

 1. Allow weighted path lengths.  This would be a very different function as one would need to specify a bin size for histogramming when path lengths are not guaranteed to be integers.

 2. Allow computing the histogram only for a subset of the start and end vertices (just like `igraph_shortest_paths()`).

Why is this useful if we already have functions to compute the distance matrices (`igraph_shortest_paths()`, `igraph_shortest_paths_dijkstra()`, etc.)?  Because `igraph_path_length_hist()` works even when the distance matrix is too large to keep in memory.  On my computer I can run the histogramming function for a 40000-vertex graph, but I cannot run the distance matrix computation without running out of memory.

**Comment by ntamas** [comment]
`igraph_shortest_paths_dijkstra()` could probably be used as a workaround until this is implemented properly - just run it with a single source vertex, for each of the vertices in your graph. This way you only have to keep a single row of the distance matrix in memory.

**Comment by heavywatal** [comment]
+1 to the second feature. In phylogenetic analysis, we often need to calculate the distances between leaf/sink vertices. The workaround that @ntamas suggested (applying `igraph_shortest_paths()` per column or row) is not so bad in terms of memory usage, but the speed is not satisfactory. E.g., It takes ~200 times longer than `igraph_path_length_hist()` for a tree with 1023 vertices. I guess construction/destruction of adjlist and some other intermediate objects cost a lot.

**Comment by szhorvat** [comment]
@heavywatal Can you provide a sample input (possibly generated) for benchmarking purposes?

**Comment by szhorvat** [comment]
(I already have Tamás's workaround implemented for IGraph/M but I see some weird performance results that I do not understand.)

**Comment by heavywatal** [comment]
@szhorvat The input is a simple tree from `igraph_tree()` with 1023 vertices and 2 children for each. `from` and `to` are the 512 leaf/sink vertices. The source code I wrote is here:
https://github.com/heavywatal/rigraphlite/blob/5cf98c3114af10e044684f754d052583584cbe62/src/structural.cpp#L84-L113

Sorry I have no time right now to show you how to run this Rcpp code for benchmarking. I will come back (at least) a couple of hours later.

**Comment by szhorvat** [comment]
Here's a quick test using IGraph/M:

![image](https://user-images.githubusercontent.com/1212871/54598494-e1e4c100-4a38-11e9-9cc5-cd34081ac339.png)
![image](https://user-images.githubusercontent.com/1212871/54598650-43a52b00-4a39-11e9-93f5-8fa8fbcbba5c.png)


What I see here is that:

 - Compared to one call to igraph_shortest_paths, vertex-by-vertex processing is considerably slower. See 0.014 s vs 0.077 seconds (5.5x slowdown). It is worth investigating if vertices should be processed in batches instead of one-by-one.

 - The slowdown is not nearly as extreme as you claim (200x).

You can check the source here: https://github.com/szhorvat/IGraphM/blob/master/IGraphM/LibraryResources/Source/IG.h#L1396

I remember that when I originally wrote this code, I investigated batch-processing vertices, but I do not remember why I gave up on it.

**Comment by szhorvat** [comment]
> The neighborhood mode is fixed to OUT in your code

I ran it on an undirected graph, so it should not matter.

**Comment by heavywatal** [comment]
You are right: the mode does not matter on undirected graphs, and it is equivalent to the ALL mode operation on directed graphs. I just noticed by reading your code that I haven't measured the directional operations on directed trees. And here is the benchmark of `shortest_path()` for each neighborhood mode:

```r
bench::mark(
  OUT___1col = g$shortest_paths(from, to[1], numeric(0), 1L, ""),
   IN___1col = g$shortest_paths(from, to[1], numeric(0), 2L, ""),
  ALL___1col = g$shortest_paths(from, to[1], numeric(0), 3L, ""),
  ALL_512col = g$shortest_paths(from, to, numeric(0), 3L, ""),
  check = FALSE
)[,1:10]
#   expression          min         mean       median          max    itr/sec     mem_alloc  n_gc n_itr   total_time
#        <chr> <bench_time> <bench_time> <bench_time> <bench_time>      <dbl> <bench_bytes> <dbl> <int> <bench_time>
# 1 OUT___1col      268.4µs      279.2µs     274.48µs     605.32µs 3581.70844        8.09KB     1  1776        496ms
# 2  IN___1col      356.3µs     366.66µs     361.79µs     541.37µs 2727.32030        8.09KB     1  1354        496ms
# 3 ALL___1col        6.9ms       7.17ms       7.06ms       7.92ms  139.38327        8.09KB     0    70        502ms
# 4 ALL_512col       12.7ms      13.22ms      13.17ms      14.03ms   75.64487        2.01MB     2    36        476ms
```

The mode matters on directed graphs, and it seams reasonable qualitatively. In OUT mode, search from each leaf ends quickly because no other vertices are reachable. In IN mode, the result is the same (all `Inf`) but it takes a bit longer to reach the root. In ALL mode, it takes very long time to reach all other vertices.

The difference between 3 and 4 is interesting. 3 returns a 512x1 matrix in ~7ms, 4 returns a 512x512 matrix in ~13ms.

**Comment by heavywatal** [comment]
No, in my environment, `igraph_shortest_paths()` on an undirected graph is as slow as `igraph_shortest_paths(..., mode=IGRAPH_ALL)` on a directed graph.

I am not familiar with pure C development and it may take some time for me to make a minimal reproducible code in C. As seen in the source code, it is just a simple wrapper of `igraph_shortest_paths()`: https://github.com/heavywatal/rigraphlite/blob/8da5532b87610e50e1f55d1f0d6137aecfc3651c/src/structural.cpp#L18-L33

**Comment by heavywatal** [comment]
Here is a minimal reproducible code:
https://github.com/heavywatal/rigraphlite/blob/master/tools/benchmark_shortest_paths.c#L1

A result in my environment is like this:
```
% clang -DNDEBUG -Wall -pedantic -O2 -I${HOME}/local/include -L${HOME}/local/lib -ligraph ./benchmark_shortest_paths.c && ./a.out
vcount: 1023, sink_size: 512, sink_first: 511
1553495563.454439000
igraph_shortest_paths [512 x 512]
1553495563.464499000
igraph_shortest_paths [512 x 1] x 512
1553495566.387414000
```
`igraph_shortest_paths [512 x 512]` takes ~10ms, while `igraph_shortest_paths [512 x 1] x 512` takes ~3s.

**Comment by heavywatal** [comment]
I tried row-by-row as well, out of curiosity. Interestingly, it takes only ~80ms and is way faster than col-by-col:

```
vcount: 1023, sink_size: 512, sink_first: 511
1553497331.511539000
igraph_shortest_paths [512 x 512]
1553497331.522315000
igraph_shortest_paths [512 x 1] x 512
1553497334.458432000
igraph_shortest_paths [1 x 512] x 512
1553497334.538309000
```

**Comment by ntamas** [comment]
In the second case (the one that takes a long time), you are basically telling igraph "okay, start a single breadth-first search from vertex X and find shortest paths to _all_ the nodes". Since your graph is a tree, you are starting from a sink and you are conducting BFS, igraph will, as a "side effect", find the shortest paths between all node pairs in order to find the shortest path from vertex X to the "farthest" node in the tree. Most of the results are then discarded because the original invocation only asked for the shortest paths that originate from X, it does not matter that you have found many other shortest paths in the process.

In the third case, you are telling igraph "okay, start multiple breadth-first searches from _all_ the nodes simultaneously, and stop as soon as you reach vertex X from all of them". I think that here we waste less effort during the BFS because many of the branches in the search tree can be cut prematurely, although I can't formulate it exactly why this happens. But the net result is that this way less effort is "wasted" by finding paths whose length will not be used in the result in the end.

I think that the general take-home message is that we could speed up the second case by doing the following:

When the "mode" parameter is "ALL", igraph should check whether the "from" or the "to" set is smaller. If the "from" set is smaller, `igraph_shortest_paths()` should be invoked recursively with "from" and "to" being swapped. Of course the resulting matrix would be transposed, so `igraph_shortest_paths()` should also transpose the shortest path matrix before returning it to the user.

**Comment by ntamas** [comment]
Before doing that, one should also benchmark whether this holds for general graphs as well and not only tree graphs and sink vertices.

**Comment by szhorvat** [comment]
@ntamas This conversation is getting long. It would be useful to sort out the details in a separate conversation (chatroom), and post only the final findings here. Would the current igraph Gitter chatroom be good for this, or better set up an "igraph-devel" chatroom? https://gitter.im/igraph/igraph

**Comment by ntamas** [comment]
Joined https://gitter.im/igraph/igraph
